File: sungwoopark95/C++/LinkedList/23_Merge_K_Sorted_Lists.cpp
#include <iostream>
#include <vector>

using namespace std;

class ListNode {
public:
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(NULL) {}
};

ListNode* create_linked_list(const vector<int>& l) {
    if (l.empty()) {
        return nullptr;
    }

    ListNode* start = new ListNode(l[0]);
    ListNode* node = start;
    for (int i = 1; i < l.size(); ++i) {
        ListNode* node_new = new ListNode(l[i]);
        node->next = node_new;
        node = node_new;
    }
    return start;
}

vector<int> linked_to_list(ListNode* head) {
    vector<int> ans;
    ListNode* curr = head;
    while (curr) {
        ans.push_back(curr->val);
        curr = curr->next;
    }
    return ans;
}

vector<int> checkNones(const vector<ListNode*>& lists) {
    vector<int> not_nones;
    for (int i = 0; i < lists.size(); ++i) {
        if (lists[i]) {
            not_nones.push_back(i);
        }
    }
    return not_nones;
}

ListNode* concatLists(vector<ListNode*>& lists, const vector<int>& indices) {
    ListNode dummy(0);
    ListNode* curr = nullptr;
    for (int i = 0; i < indices.size(); ++i) {
        int idx = indices[i];
        ListNode* head = lists[idx];
        if (i == 0) {
            dummy.next = head;
            curr = dummy.next;
        } else {
            curr->next = head;
            curr = head;
        }
        while (curr && curr->next) {
            curr = curr->next;
        }
    }
    return dummy.next;
}

ListNode* insert_sort_list(ListNode* head) {
    ListNode dummy(0);
    dummy.next = head;
    ListNode* prev = &dummy;
    ListNode* curr = head;

    while (curr) {
        if (curr->next && curr->val > curr->next->val) {
            while (prev->next && prev->next->val < curr->next->val) {
                prev = prev->next;
            }

            ListNode* temp = curr->next;
            curr->next = temp->next;
            temp->next = prev->next;
            prev->next = temp;
            prev = &dummy;
        } else {
            curr = curr->next;
        }
    }
    return dummy.next;
}

ListNode* mergeKLists(vector<ListNode*>& lists) {
    vector<int> valid_index = checkNones(lists);
    if (valid_index.empty()) {
        return nullptr;
    }
    ListNode* head = concatLists(lists, valid_index);
    return insert_sort_list(head);
}

int main() {
    vector<vector<int>> lists = {{1, 4, 5}, {1, 3, 4}, {2, 6}};
    vector<ListNode*> links;

    for (const auto& item : lists) {
        links.push_back(create_linked_list(item));
    }

    ListNode* merged = mergeKLists(links);

    vector<int> result = linked_to_list(merged);
    for (int val : result) {
        cout << val << " ";
    }
    cout << endl;

    // 메모리 정리
    for (ListNode* node : links) {
        while (node) {
            ListNode* temp = node;
            node = node->next;
            delete temp;
        }
    }
}

File: sungwoopark95/python/LinkedList/23_Merge_K_Sorted_Lists.py
from typing import List

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

# Create a linked list from a list, then return head node
def create_linked_list(l: list) -> ListNode:
    if len(l) == 0:
        return None
    
    start = ListNode(l[0])
    node = start
    
    for i in range(1, len(l)):
        node_new = ListNode(l[i])
        node.next = node_new
        node = node_new
    return start

def linked_to_list(head:ListNode) -> list:
    ans = []
    curr = head
    while curr:
        ans.append(curr.val)
        curr = curr.next
    return ans

def checkNones(lists) -> List[int]:
    not_nones = []
    for i, node in enumerate(lists):
        if not node:
            continue
        not_nones.append(i)
    return not_nones

def concatLists(lists: List[ListNode], indices: List[int]) -> ListNode:
    # dummy는 연결 리스트를 편리하게 다루기 위한 더미 노드입니다.
    dummy = ListNode()
    
    # curr는 현재 연결 리스트의 마지막 노드를 가리키는 포인터입니다.
    curr = None
    
    # 주어진 인덱스의 순서에 따라 각 연결 리스트를 연결합니다.
    for i in range(len(indices)):
        # 현재 연결할 리스트의 인덱스를 가져옵니다.
        idx = indices[i]
        
        # 해당 인덱스의 연결 리스트의 헤드를 가져옵니다.
        head = lists[idx]
        
        # 첫 번째 연결 리스트인 경우 dummy의 next를 첫 번째 연결 리스트의 헤드로 설정합니다.
        if i == 0:
            dummy.next = head
            curr = dummy.next
        else:
            # 그렇지 않은 경우, 현재 연결 리스트의 끝에 새로운 연결 리스트를 연결합니다.
            curr.next = head
            curr = head
        
        # curr를 현재 연결 리스트의 마지막 노드로 이동시킵니다.
        while curr.next:
            next = curr.next
            curr = next
            
    # 더미 노드 다음 노드(실제 연결 리스트의 첫 번째 노드)를 반환합니다.
    return dummy.next

def insert_sort_list(head:ListNode):
    dummy = ListNode()
    dummy.next = head # dummy node를 head에 연결
    prev = dummy # dummy node에 Prev 포인트
    
    curr = head # head node에 curr 포인트
    # next = curr.next
    
    # curr node만 움직이면서 보기 
    while curr: 
        # 만약 비교할 Curr.next 값이 있고 현재 값이 다음 값보다 큰 경우 (swap)
        # 4 > 2
        if curr.next and curr.val > curr.next.val:
            # insert할 위치 찾기 (prev 이동시키면서)
            # prev.next.val = 4 > curr.next = 2 (while 스킵)
            while prev.next.val < curr.next.val:
                prev = prev.next
            
            # temp = curr.next = 2
            temp = curr.next # curr.next를 temp로 지정해서 풀기
            curr.next = temp.next # 4 다음 노드 -> temp.next = 1
            temp.next = prev.next # 2 다음 노드 -> prev 다음 = 4
            prev.next = temp        # prev 다음 노드 -> temp = 2
        
            prev = dummy # dummy로 prev 보내기 (바꾸고 역할 끝)
        else:
            curr = curr.next # curr을 다음으로 넘기기
        # next = curr.next
    return prev.next # dummy.next

def mergeKLists(lists: List[ListNode]) -> ListNode:
    valid_index = checkNones(lists) # None이 아닌 node들의 index로 이루어진 list
    # 모든 node가 None으로만 이루어져 있다면 None return
    if not valid_index:
        return None
    head = concatLists(lists, valid_index)
    return insert_sort_list(head)

if __name__ == "__main__":
    lists = [[1,4,5],[1,3,4],[2,6]]
    links = []
    for item in lists:
        links.append(create_linked_list(item))
    merged = mergeKLists(links)
    print(linked_to_list(merged))